File: top150/ArrayString/q380.java
package top150.ArrayString;

import java.util.HashMap;
import java.util.Map;
import java.util.Random;

public class q380 {
    class RandomizedSet {

        Map<Integer, Integer> map;
        Random rand;
        int[] arr;
        int size;
        

        public RandomizedSet() {
            map = new HashMap<Integer, Integer>();
            rand = new Random();
            arr = new int[10_000];
            size = 0;
        }
        
        public boolean insert(int val) {
            if(map.containsKey(val)) {
                return false;
            } else {
                map.put(val, size);
                arr[size] = val;
                size++;
                return true;
            }
        }
        
        public boolean remove(int val) {
            if(map.containsKey(val)) {
                int valIndex = map.get(val);
                int temp = arr[size-1];
                arr[valIndex] = temp;
                map.put(temp, valIndex);
                map.remove(val);
                size--;
                return true;
            } else {
                return false;
            }
        }
        
        public int getRandom() {
            // System.out.println(Arrays.toString(arr));
            int r = rand.nextInt(size);
            return arr[r];
        }
    }
}

File: top150/ArrayString/q380.py
from collections import defaultdict
import random


class RandomizedSet:

    def __init__(self):
        self.d = defaultdict(int)
        self.l = []

    # T: O(1), S: O(N), N: length of values in self.l/self.d
    def insert(self, val: int) -> bool:
        if val not in self.d.keys():
            self.l.append(val)
            self.d[val] = len(self.l) - 1
            return True
        else:
            return False

    # T: O(1), S: O(N), N: length of values in self.l/self.d
    def remove(self, val: int) -> bool:
        if val not in self.d.keys():
            return False
        else:
            idx = self.d[val]
            if idx != len(self.l) - 1:
                last_element = self.l[-1]
                self.l[idx] = last_element
                self.d[last_element] = idx
            self.l.pop()
            del self.d[val]
            return True

    # T: O(1), S: O(N), N: length of values in self.l/self.d
    def getRandom(self) -> int:
        return random.choice(self.l)
